test/test_testing.py

```diff
@@ -4463,6 +4333,102 @@ def test_make_dir_with_permission_mask(self):
         pass
 
 
+class TestFilesystem(unittest.TestCase, _TestingPebbleClientMixin):
+    def setUp(self) -> None:
+        self.harness = ops.testing.Harness(ops.CharmBase, meta='''
+            name: test
+            containers:
+                test-container:
+                    mounts:
+                        - storage: test-storage
+                          location: /mounts/foo
+            storage:
+                test-storage:
+                    type: filesystem
+            ''')
+        self.harness.begin()
+        self.harness.set_can_connect("test-container", True)
+        self.root = self.harness.get_filesystem_root("test-container")
+        self.container = self.harness.charm.unit.get_container("test-container")
+
+    def tearDown(self) -> None:
+        self.harness.cleanup()
+
+    def test_push(self):
+        self.container.push("/foo", source="foo")
+        self.assertTrue((self.root / "foo").is_file())
+        self.assertEqual((self.root / "foo").read_text(), "foo")
+
+    def test_push_create_parent(self):
+        self.container.push("/foo/bar", source="bar", make_dirs=True)
+        self.assertTrue((self.root / "foo").is_dir())
+        self.assertEqual((self.root / "foo" / "bar").read_text(), "bar")
+
+    def test_push_path(self):
+        with tempfile.TemporaryDirectory() as temp:
+            tempdir = pathlib.Path(temp)
+            (tempdir / "foo/bar").mkdir(parents=True)
+            (tempdir / "foo/test").write_text("test")
+            (tempdir / "foo/bar/foobar").write_text("foobar")
+            self.container.push_path(tempdir / "foo", "/tmp")
+
+            self.assertTrue((self.root / "tmp").is_dir())
+            self.assertTrue((self.root / "tmp/foo").is_dir())
+            self.assertTrue((self.root / "tmp/foo/bar").is_dir())
+            self.assertEqual((self.root / "tmp/foo/test").read_text(), "test")
+            self.assertEqual((self.root / "tmp/foo/bar/foobar").read_text(), "foobar")
+
+    def test_make_dir(self):
+        self.container.make_dir("/tmp")
+        self.assertTrue((self.root / "tmp").is_dir())
+        self.container.make_dir("/foo/bar/foobar", make_parents=True)
+        self.assertTrue((self.root / "foo/bar/foobar").is_dir())
+
+    def test_pull(self):
+        (self.root / "foo").write_text("foo")
+        self.assertEqual(self.container.pull("/foo").read(), "foo")
+
+    def test_pull_path(self):
+        (self.root / "foo").mkdir()
+        (self.root / "foo/bar").write_text("bar")
+        # TODO: pull_path doesn't pull empty directories, intended?
```

Comment:
Yeah, if that's the case, that seems like a bug. Can you please open a separate canonical/operator issue for that?